src/mds/CDir.cc

```diff
@@ -1815,8 +1810,18 @@ void CDir::_omap_fetched(bufferlist& hdrbl, map<string, bufferlist>& omap,
       cache->mds->clog->warn() << "Corrupt dentry '" << dname << "' in "
                                   "dir frag " << dirfrag() << ": "
                                << err;
-      go_bad();
-      return;
+
+      // Remember that this dentry is damaged.  Subsequent operations
+      // that try to act directly on it will get their EIOs, but this
+      // dirfrag as a whole will continue to look okay (minus the
+      // mysteriously-missing dentry)
+      go_bad_dentry(last, dname);
+
+      // Anyone who was WAIT_DENTRY for this guy will get kicked
+      // to RetryRequest, and hit the DamageTable-interrogating path.
+      // Stats will now be bogus because we will think we're complete,
+      // but have 1 or more missing dentries.
+      continue;
```

Comment:
If we don't mark the directory as complete, then other callers will get stuck: the behaviour I'm going for here is that a damaged dentry disappears, but callers can still see other dentries in the directory.

The alternative would be to make the whole directory inaccessible when we see a bad dentry (pretty much the old behaviour)